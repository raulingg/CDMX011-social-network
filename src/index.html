<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <script src="https://kit.fontawesome.com/eb71235840.js" crossorigin="anonymous"></script>
  <link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Bebas+Neue&family=Roboto&display=swap" rel="stylesheet">

  <link rel="stylesheet" href="./style.css">
  <title>Ã‰chale la culpa</title>
</head>
<body>
  <div id="root"></div>


  <!-- The core Firebase JS SDK is always required and must be listed first -->
<script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-app.js"></script>

<!-- TODO: Add SDKs for Firebase products that you want to use
     https://firebase.google.com/docs/web/setup#available-libraries -->
     <script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-firestore.js"></script>

     <script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-auth.js"></script>


  <script type="module" src="main.js"></script>
  <script type="module" src="./lib/firebase.js"></script>

</body>
</html>