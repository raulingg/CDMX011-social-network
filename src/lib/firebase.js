/* eslint-disable import/no-cycle */
// import { onNavigate } from '../router/routes.js';

/* eslint-disable no-unused-vars */
const firebaseConfig = {
  apiKey: 'AIzaSyDhl8UbwLZqgbT0ZkSZnECYAijJ6cq-3Yw',
  authDomain: 'echale-la-culpa.firebaseapp.com',
  projectId: 'echale-la-culpa',
  storageBucket: 'echale-la-culpa.appspot.com',
  messagingSenderId: '935955315781',
  appId: '1:935955315781:web:d3f1c617ea7b2e7dec372c',
};
// Initialize Firebase
firebase.initializeApp(firebaseConfig);
const db = firebase.firestore();

const provider = new firebase.auth.GoogleAuthProvider();

export const login = (email, password, onNavigate) => {
  firebase.auth().signInWithEmailAndPassword(email, password)
    .then((userCredential) => {
    // Signed in
      const user = userCredential.user;
      console.log(user);
      onNavigate('/timeline');
    // ...
    })
    .catch((error) => {
      const errorCode = error.code;
      const errorMessage = error.message;
    });
};
export const observer = async () => {
  const user = await firebase.auth().currentUser;
  const userUid = user.uid;
  return userUid;
};
export const createUsser = (email, password) => {
  firebase
    .auth()
    .createUserWithEmailAndPassword(email, password)
    .then((userCredential) => {
      // Signed in
      console.log(userCredential);
      const user = userCredential.user;
      // ...
    })
    .catch((error) => {
      const errorCode = error.code;
      const errorMessage = error.message;
      alert(`oh no! ${errorMessage}`);
      console.log(errorCode);
      // ..
    });
};

export const publishPost = (singer, song) => {
  db.collection('post').doc().set({
    singer,
    song,
    likes: [],
  })
    .then(() => {
      console.log('post creado');
    })
    .catch((error) => {
      console.error('Buu no se creÃ³ el post', error);
    });
};
export const deletePost = (id) => db.collection('post').doc(id).delete();

export const getAllPost = (callback) => {
  db.collection('post').onSnapshot(callback);
  /* querySnapshot.docs.forEach((doc) => {
      const post = doc.data();
      post.id = doc.id;
      posts.push(post);
    }); */
};

export const continueWithGoogle = (onNavigate) => {
  firebase.auth()
    .signInWithPopup(provider)
    .then((result) => {
    /** @type {firebase.auth.OAuthCredential} */
      const credential = result.credential;

      // This gives you a Google Access Token. You can use it to access the Google API.
      const token = credential.accessToken;
      // The signed-in user info.
      const user = result.user;
      onNavigate('/timeline');
    // ...
    }).catch((error) => {
    // Handle Errors here.
      const errorCode = error.code;
      const errorMessage = error.message;
      // The email of the user's account used.
      const email = error.email;
      // The firebase.auth.AuthCredential type that was used.
      const credential = error.credential;
    // ...
    });
};

export const updateLike = (id, uid) => db.collection('post').doc(id).update({
  likes: firebase.firestore.FieldValue.arrayUnion(uid),
});
export const updateUnLike = (id, uid) => db.collection('post').doc(id).update({
  likes: firebase.firestore.FieldValue.arrayRemove(uid),
});

export const signOut = (onNavigate) => {
  firebase.auth().signOut().then(() => {
    onNavigate('/');
  }).catch((error) => {
    // An error happened.
  });
};
