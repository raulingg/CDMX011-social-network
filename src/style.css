body {
    background-color: black;
    margin: 0;
    padding: 0;
    font-family: 'Roboto', sans-serif;
}
#root{
    margin: 0;
    padding: 0;
}
/* Estilos home */
#homeContainer{
    color: #fff;
    width: 100%;
    height: 100vh;
    display: flex;
    flex-direction: column;
    background: url('./images/cover-home.jpg') no-repeat center;
    background-size: cover;
    margin: 0;
    padding: 0;
    justify-content: center;
    font-family: 'Gill Sans', sans-serif;
    text-align: center;

}
#divRegister {
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background: url('./images/cover-signup.jpg') no-repeat center;
    background-size: cover;
}
#divRegister form {
    display: flex;
    width: 50%;
    height: 100px;
    flex-direction: column;
    justify-content: space-between;
}
#divLogin {
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background: url('./images/cover-login.jpg') no-repeat center;
    background-size: cover;

}
#divLogin form {
    display: flex;
    width: 50%;
    height: 100px;
    flex-direction: column;
    justify-content: space-between;
}
input:invalid {
    border: 2px solid red;
}
/* Time line */
.container {
    min-height: 100vh;
    position: relative;
    width: 100%;
    color: white;
}

#menu {
    position: fixed;
    bottom: 0;
    background-color: #18BCB7;
    width: 100%;
    height: 100px;
    display: flex;
}
#menu button{
    height: 100%;
    width: 33.33%;
    margin: 0;
    background-color: #18BCB7;
    border: none;
    border-right: 1px solid white;
    transition: all .5s;
}
#menu button:hover{
background-color:#118d89 ;
} 
#menu button:hover i{
transform: rotate(10deg);
} 
#menu button i{
    font-size: 1.5rem;
    color: white;
}
.create-post{
    display: flex;
    text-align: center;
    justify-content: center;
}
#postForm {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: auto;
    text-align: left;
}
label {
    margin-bottom: 10px;
}
#singer{
    margin-bottom: 30px;
}
#song{
    margin-bottom: 30px;
}
#publish{
    background-color: #FAA138;
    color: #fff;
    border: none;
    padding: 10px;
    transition: all .5s;
    letter-spacing: 0;
}
#publish:hover{
    background-color: #c97b22;
    letter-spacing: 2px;
}

/*     font-family: 'Bebas Neue', cursive;
 */

 /* Cards */
 .timeline {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: 100px 0;
}
#timeLineContainer{
    position: relative;
    background: url('./images/cover-timeline.jpg') no-repeat center;
    background-size: cover;
    background-attachment: fixed;

}
#timeLineContainer::before{
    content: '';
    background-color: black;
    opacity: .8;
    width: 100%;
    height: 100%;
    position: absolute;
}

 #timeLineContainer h1,h2,h3{
     padding: 0;
     margin: 0;
 }
 .card{
     width: 50%;
     text-align: center;
     border: 1px solid #fff;
     margin-bottom: 30px;
     text-transform: uppercase;
     position: relative;
     padding: 0px 20px 30px 20px;
     display: flex;
     flex-direction: column;
     align-items: center;
 }
 .header-card{
     position: absolute;
     background-color: #FAA138;
     padding: 5px 20px;
     width: auto;
     min-width: 50%;
     display: block;
     float: left;
     border-radius: 20px;
     left: 50%;
     transform: translateX(-50%);
     margin-top: -15px;
     margin-bottom: 30px;
 }
 .singer-name{
     margin-top: 50px;
 }
 .song-sentence{
    font-family: 'Bebas Neue', cursive;
    -webkit-text-stroke: .25px rgb(226, 224, 224);
    color: rgba(0, 0, 0, .5); 
    font-size: 4rem;
 }
 .options-post{
     margin-top: 20px;
     width: 50%;
     display: flex;
     justify-content: space-between;
 }
 .card i{
     cursor: pointer;
 }
 .card i:hover{
     color: #FAA138;
 }