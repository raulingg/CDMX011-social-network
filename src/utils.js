export function render(target, element) {
  while (target.firstChild) {
    target.removeChild(target.lastChild);
  }

  target.appendChild(element);
}
