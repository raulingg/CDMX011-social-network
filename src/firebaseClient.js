import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';

const firebaseConfig = {
  apiKey: 'AIzaSyAxKjqTudxBWi9j9i4IDKn450hmM1m_w3A',
  authDomain: 'red-social-coaches.firebaseapp.com',
  projectId: 'red-social-coaches',
  appId: '1:489293961166:web:e0ae10e7b6db857d26fd9f',
};

firebase.initializeApp(firebaseConfig);

export const signIn = (email, password) => firebase.auth()
  .signInWithEmailAndPassword(email, password);

export const signOut = () => firebase.auth().signOut();

export const getUser = () => firebase.auth().currentUser;
