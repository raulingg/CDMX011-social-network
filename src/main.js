import { dispatchRoute } from './routes';

window.addEventListener('load', () => {
  dispatchRoute('/');
});

window.addEventListener('popstate', () => {
  dispatchRoute(window.location.pathname);
});
