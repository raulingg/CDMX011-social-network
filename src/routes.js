import { Login } from './views/Login.js';
import { Home } from './views/Home.js';
import { render } from './utils.js';

const routes = {
  '/wall': Home,
  '/': Login,
};

export const dispatchRoute = (pathname = '/') => {
  const root = document.getElementById('root');
  const component = routes[pathname];
  render(root, component());
};

export const onNavigate = (pathname) => {
  window.history.pushState({}, pathname, window.location.origin + pathname);
  dispatchRoute(pathname);
};
