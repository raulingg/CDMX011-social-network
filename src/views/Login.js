import { signIn } from '../firebaseClient.js';
import { onNavigate } from '../routes.js';

export const Login = () => {
  const container = document.createElement('div');

  const html = `
    <form id="signInForm">
        <label for="email">E-mail</label>
        <input id="email" type="email">
        <label for="password">Password</label>
        <input id="password" type="password">
        <input id="signInBtn" type="submit" value="Sign In">
        <p id="errorMessage"></p>
    </form>
    `;

  container.innerHTML = html;

  container.querySelector('#signInBtn').addEventListener('click', (event) => {
    event.preventDefault();

    const email = container.querySelector('input[type=email]').value;
    const password = container.querySelector('input[type=password]').value;

    signIn(email, password)
      .then(() => onNavigate('/wall'))
      .catch((error) => {
        container.querySelector('#errorMessage').innerHTML = error.message;
      });
  });

  return container;
};
