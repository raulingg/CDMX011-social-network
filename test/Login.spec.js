/**
 * @jest-environment jsdom
 */
import './globals/firebase';
import { render } from '../src/utils';
import { Login } from '../src/views/Login';

const delay = (timeInMs = 100) => new Promise((resolve) => setTimeout(resolve, timeInMs));

describe('Login', () => {
  document.body.innerHTML = '<div id="root"></div>';
  const rootDiv = document.getElementById('root');

  const fillOutAndSubmitForm = (email, password) => {
    document.getElementById('email').value = email;
    document.getElementById('password').value = password;
    document.getElementById('signInBtn').click();
  };

  it('renderiza', () => {
    render(rootDiv, Login());

    // ref: https://jestjs.io/es-ES/docs/snapshot-testing
    expect(rootDiv.innerHTML).toMatchSnapshot();
  });

  it('login al usuario exitosamente', async () => {
    const email = 'test@laboratoria.la';
    const password = '123456';
    const user = { email, uid: 'xxxxxxx' }; // un objeto dummy que representa el user

    // Generamos una promesa resuelta (exitosa) para este escenario, la cual retornara
    // nuestro objeto user dummy
    // ref: https://jestjs.io/docs/mock-function-api#mockfnmockresolvedvaluevalue
    const mockSignInWithEmailAndPassword = jest.fn();
    mockSignInWithEmailAndPassword.mockResolvedValue(user);

    // mockeamos solo los métodos necesarios de firebase.auth() para este test
    const mockFirebaseAuth = {
      signInWithEmailAndPassword: mockSignInWithEmailAndPassword,
      currentUser: user,
    };

    // La clave firebase.auth ahora sera una función que retorna un objeto con los mocks
    // cuando firebase.auth() sea invocado este retornara nuestro objeto mock mockFirebaseAuth
    firebase.auth = () => mockFirebaseAuth;

    render(rootDiv, Login());
    fillOutAndSubmitForm(email, password);

    // diferimos 100 ms la ejecución de los expects, debido a que los eventos del DOM
    // son asíncronos y no podemos saber en que momento terminara la resolución de nuestra
    // promesa signInWithEmailAndPassword. Al usar promesas dentro de nuestro test
    // nos vemos en la necesidad de hacer nuestro test asíncrono.
    // ref: https://jestjs.io/docs/asynchronous#promises
    await delay(100);

    // Comprobamos que nuestro mock del método firebase.auth().signInWithEmailAndPassword()
    // fue invocado con el email y password que usamos en el llenado del form
    // ref: https://jestjs.io/docs/next/expect#tohavebeencalledwitharg1-arg2-
    expect(mockSignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);

    // Comprobamos que en el contenido de la pagina ahora aparece el email del usuario logeado
    // ref: https://jestjs.io/docs/next/expect#tocontainitem
    expect(rootDiv.innerHTML).toContain(email);
  });

  it('muestra un mensaje de error cuando las credenciales son invalidas', async () => {
    const email = 'test@laboratoria.la';
    const password = 'invalidpassword';
    const errorMessage = 'Invalid credentials!';

    // En este caso generamos una promesa rechazada que nos devuelve el error
    // ref: https://jestjs.io/docs/mock-function-api#mockfnmockrejectedvaluevalue
    const mockSignInWithEmailAndPassword = jest.fn();
    mockSignInWithEmailAndPassword.mockRejectedValue(new Error(errorMessage));

    const mockFirebaseAuth = {
      signInWithEmailAndPassword: mockSignInWithEmailAndPassword,
    };

    firebase.auth = () => mockFirebaseAuth;

    render(rootDiv, Login());
    fillOutAndSubmitForm(email, password);

    await delay(100);

    expect(mockSignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);

    // Verifica si dentro del contenido HTML se encuentra el mensaje de error
    expect(rootDiv.innerHTML).toContain(errorMessage);
  });
});
