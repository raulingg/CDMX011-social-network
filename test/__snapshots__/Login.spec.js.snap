// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renderiza 1`] = `
"<div>
    <form id=\\"signInForm\\">
        <label for=\\"email\\">E-mail</label>
        <input id=\\"email\\" type=\\"email\\">
        <label for=\\"password\\">Password</label>
        <input id=\\"password\\" type=\\"password\\">
        <input id=\\"signInBtn\\" type=\\"submit\\" value=\\"Sign In\\">
        <p id=\\"errorMessage\\"></p>
    </form>
    </div>"
`;
